### PR TITLE
[IMP] PieChart: percentage visualization for tooltip

### PR DESCRIPTION
**Pull request by aath-odoo**
## Description:

The pie chart tooltips have been enhanced to show the calculated percentages along with their corresponding dataset values. Furthermore, the labels on the chart have been updated to accurately reflect the corresponding dataset information.

Task: : [3389036](https://www.odoo.com/web#id=3389036&cids=2&menu_id=6478&action=4043&model=project.task&view_type=form)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: tests/plugins/chart/basic_chart.test.ts [hunk 1]
```diff
@@ -1527,24 +1527,24 @@ describe("Chart design configuration", () => {
     expect(chart.chartJsConfig.data!.labels).toEqual(["3/1/2022", "2022/03/02"]);
   });
 
+  function getTooltipItem(value: number, chartType: string, chart): TooltipItem<ChartType> {
```
Comment (somu-odoo) [inline tests/plugins/chart/basic_chart.test.ts:1530]: Add a type for `chart`

File: tests/plugins/chart/basic_chart.test.ts [hunk 2]
```diff
@@ -1527,24 +1527,24 @@ describe("Chart design configuration", () => {
...
+    return {
+      chart: chart,
```
Comment (somu-odoo) [inline tests/plugins/chart/basic_chart.test.ts:1532]:
```suggestion
      chart,
```

**Review by somu-odoo — CHANGES_REQUESTED** [review]
Couple minor changes, also can you please add a test case for negative dataset.

File: src/helpers/figures/charts/pie_chart.ts [hunk 3]
```diff
@@ -212,6 +213,23 @@ function getPieConfiguration(
   config.options.plugins!.tooltip!.callbacks!.title = function (tooltipItems) {
     return tooltipItems[0].dataset.label;
   };
+  config.options.plugins!.tooltip!.callbacks!.label = function (tooltipItems) {
```
Comment (hokolomopo) [inline src/helpers/figures/charts/pie_chart.ts:216]:
```suggestion
  config.options.plugins!.tooltip!.callbacks!.label = function (tooltipItem) {
```

File: src/helpers/figures/charts/pie_chart.ts [hunk 4]
```diff
@@ -225,6 +243,24 @@ function getPieColors(colors: ChartColors, dataSetsValues: DatasetValues[]): Col
   return pieColors;
 }
 
+function calculatePercentageValues(dataSetsValues) {
```
Comment (hokolomopo) [inline src/helpers/figures/charts/pie_chart.ts:246]: Please type the argument and return values (even is using ChartJs types in annoying). Everything is typed as any, so you have some wrong types and nobody noticed. (data.map returning a single string? that's not how `map` works...)

And As stated in another comment, we can simplify _a lot_ the function to take a single dataset as argument, and return the percentage for a single value

File: src/helpers/figures/charts/pie_chart.ts [hunk 5]
```diff
@@ -212,6 +213,23 @@ function getPieConfiguration(
   config.options.plugins!.tooltip!.callbacks!.title = function (tooltipItems) {
     return tooltipItems[0].dataset.label;
   };
+  config.options.plugins!.tooltip!.callbacks!.label = function (tooltipItems) {
+    const { format, locale } = localeFormat;
+    const data = tooltipItems.chart.data;
+    const percentageValues = calculatePercentageValues(data.datasets);
```
Comment (hokolomopo) [inline src/helpers/figures/charts/pie_chart.ts:219]: Use `tooltipItems.dataSet` maybe ? And what's the point of computing the percentages values for both datasets ? We only need the valeus for the current dataset. And we only need the value of the current `dataIndex`, we don't care about the others, it's useless to compute them.

File: tests/plugins/chart/basic_chart.test.ts [hunk 6]
```diff
@@ -1527,24 +1527,29 @@ describe("Chart design configuration", () => {
     expect(chart.chartJsConfig.data!.labels).toEqual(["3/1/2022", "2022/03/02"]);
   });
 
+  function getTooltipItem(
+    value: number,
+    chartType: string,
+    chart: ChartConfiguration
```
Comment (hokolomopo) [inline tests/plugins/chart/basic_chart.test.ts:1533]: Why did you add `chart` as an argument, if it for doing nothing with it and returning it as is ?

File: tests/plugins/chart/basic_chart.test.ts [hunk 7]
```diff
@@ -1588,52 +1593,113 @@ describe("Chart design configuration", () => {
...
+        chartType,
+        runtime.chartJsConfig
+      );
+      expect(
+        // @ts-ignore should be binded to the chart tooltip model
+        runtime.chartJsConfig.options!.plugins!.tooltip!.callbacks!.label!(tooltipItem)
+      ).toEqual("6,000");
+    });
+  });
+  describe("Pie Chart tooltip", () => {
```
Comment (hokolomopo) [inline tests/plugins/chart/basic_chart.test.ts:1649]: add a newline between tests

File: tests/plugins/chart/basic_chart.test.ts [hunk 8]
```diff
@@ -1588,52 +1593,113 @@ describe("Chart design configuration", () => {
...
+        chart.chartJsConfig
+      );
+      const expectedLabel = "60,000,000 (100.00%)";
+
+      expect(
+        // @ts-ignore
+        chart.chartJsConfig!.options!.plugins!.tooltip!.callbacks!.label!(tooltipItem)
+      ).toBe(expectedLabel);
+    });
+    test("pie chart tooltip label with format", () => {
```
Comment (hokolomopo) [inline tests/plugins/chart/basic_chart.test.ts:1676]: newline

File: src/helpers/figures/charts/pie_chart.ts [hunk 9]
```diff
@@ -212,6 +213,23 @@ function getPieConfiguration(
...
+    const dataIndex = tooltipItems.dataIndex;
+    const datasetIndex = tooltipItems.datasetIndex;
+    const percentage = percentageValues[datasetIndex || 0][dataIndex!];
+
+    let xLabel = tooltipItems.label;
+    const yLabel = tooltipItems.parsed.y ?? tooltipItems.parsed;
+    const yLabelStr =
+      format && typeof yLabel === "number"
+        ? formatValue(yLabel, { format, locale })
+        : yLabel?.toLocaleString() || "";
```
Comment (hokolomopo) [inline src/helpers/figures/charts/pie_chart.ts:229]: Don't use `toLocaleString`. You can use an undefined format and non number values with `formatValue`.

File: src/helpers/figures/charts/pie_chart.ts [hunk 10]
```diff
@@ -212,6 +213,23 @@ function getPieConfiguration(
...
+    const percentageValues = calculatePercentageValues(data.datasets);
+    const dataIndex = tooltipItems.dataIndex;
+    const datasetIndex = tooltipItems.datasetIndex;
+    const percentage = percentageValues[datasetIndex || 0][dataIndex!];
+
+    let xLabel = tooltipItems.label;
+    const yLabel = tooltipItems.parsed.y ?? tooltipItems.parsed;
+    const yLabelStr =
+      format && typeof yLabel === "number"
+        ? formatValue(yLabel, { format, locale })
```
Comment (hokolomopo) [inline src/helpers/figures/charts/pie_chart.ts:228]:
```suggestion
        ? formatValue(yLabel, localeFormat)
```

File: tests/plugins/chart/basic_chart.test.ts [hunk 11]
```diff
@@ -1588,52 +1593,113 @@ describe("Chart design configuration", () => {
...
+    });
+  });
+  describe("Pie Chart tooltip", () => {
+    test("pie chart tooltip label to include percentage", () => {
+      setCellContent(model, "A1", "P1");
+      setCellContent(model, "A2", "60000000");
+      createChart(
+        model,
+        {
+          dataSets: ["A1:A2"],
```
Comment (hokolomopo) [inline tests/plugins/chart/basic_chart.test.ts:1656]: I'd add a cell with another number value in the dataset, to check that the percentage is correclty computed. (no need to test both tooltips)

File: tests/plugins/chart/basic_chart.test.ts [hunk 12]
```diff
@@ -1588,52 +1593,113 @@ describe("Chart design configuration", () => {
...
+        data.datasets![0].data![0] as number,
+        "pie",
+        chart.chartJsConfig
+      );
+      const expectedLabel = "60,000,000 (100.00%)";
+
+      expect(
+        // @ts-ignore
+        chart.chartJsConfig!.options!.plugins!.tooltip!.callbacks!.label!(tooltipItem)
+      ).toBe(expectedLabel);
```
Comment (hokolomopo) [inline tests/plugins/chart/basic_chart.test.ts:1674]: It would improve the readability to replace the `getTooltipItem` helper by an helper that takes the chart runtime, a dataset index (if we use another dataSet than 0 in some tests), and a value index, and that directly returns the result of the `label()` callback. 

It would avoid having to compute a tooltipItem that we don't care about and to write a ts-ignore in each test.

You can change all the tests that use getTooltipItem

File: tests/plugins/chart/basic_chart.test.ts [hunk 13]
```diff
@@ -1588,52 +1593,113 @@ describe("Chart design configuration", () => {
...
+        "1"
+      );
+      const chart = model.getters.getChartRuntime("1") as PieChartRuntime;
+      const data = chart.chartJsConfig.data!;
+      const tooltipItem = getTooltipItem(
+        data.datasets![0].data![0] as number,
+        "pie",
+        chart.chartJsConfig
+      );
+      const expectedLabel = "60,000,000 (100.00%)";
```
Comment (hokolomopo) [inline tests/plugins/chart/basic_chart.test.ts:1669]: The variable is kinda useless. Just put the value in the expect.

File: tests/plugins/chart/basic_chart.test.ts [hunk 14]
```diff
@@ -1588,52 +1593,113 @@ describe("Chart design configuration", () => {
...
+      setCellContent(model, "A1", "P1");
+      setCellContent(model, "A2", "60000000");
+      createChart(
+        model,
+        {
+          dataSets: ["A1:A2"],
+          labelRange: "A1",
+          type: "pie",
+        },
+        "1"
```
Comment (hokolomopo) [inline tests/plugins/chart/basic_chart.test.ts:1660]:
```suggestion
```
nitpick: we do'nt care about the ID

Comment (aath-odoo) [reply tests/plugins/chart/basic_chart.test.ts:1657]: > nitpick: we do'nt care about the ID

I guess we need Id for getChartRuntime(figureId) in order to get chart.

File: src/helpers/figures/charts/pie_chart.ts [hunk 15]
```diff
@@ -212,6 +219,19 @@ function getPieConfiguration(
   config.options.plugins!.tooltip!.callbacks!.title = function (tooltipItems) {
     return tooltipItems[0].dataset.label;
   };
+  config.options.plugins!.tooltip!.callbacks!.label = function (tooltipItem) {
+    const { format, locale } = localeFormat;
+    const data = tooltipItem.dataset?.data;
```
Comment (hokolomopo) [inline src/helpers/figures/charts/pie_chart.ts:224]:
```suggestion
    const data = tooltipItem.dataset.data;
```

File: src/helpers/figures/charts/pie_chart.ts [hunk 16]
```diff
@@ -212,6 +219,19 @@ function getPieConfiguration(
...
+    const dataIndex = tooltipItem.dataIndex;
+    const percentage = calculatePercentage(data, dataIndex);
+
+    const xLabel = tooltipItem.label || tooltipItem.dataset?.label;
```
Comment (hokolomopo) [inline src/helpers/figures/charts/pie_chart.ts:228]:
```suggestion
    const xLabel = tooltipItem.label || tooltipItem.dataset.label;
```

File: tests/plugins/chart/basic_chart.test.ts [hunk 17]
```diff
@@ -1527,24 +1534,37 @@ describe("Chart design configuration", () => {
...
+    dataset: ChartDataset<
+      keyof ChartTypeRegistry,
+      (number | [number, number] | Point | BubbleDataPoint | null)[]
+    >,
+    datasetIndex: number,
+    dataIndex: number
+  ): string {
+    const tooltipItem: TooltipItem<ChartType> = {
+      label: "",
+      // yLabel is an number for pie charts for whatever reason
```
Comment (hokolomopo) [inline tests/plugins/chart/basic_chart.test.ts:1548]:
```suggestion
```
probably an old comment that you can remove whild you're at it, I don't think it makes much sense here anymore ?

File: src/helpers/figures/charts/pie_chart.ts [hunk 18]
```diff
@@ -225,6 +245,21 @@ function getPieColors(colors: ChartColors, dataSetsValues: DatasetValues[]): Col
   return pieColors;
 }
 
+function calculatePercentage(
+  dataset: (number | [number, number] | Point | BubbleDataPoint | null)[],
```
Comment (hokolomopo) [inline src/helpers/figures/charts/pie_chart.ts:249]: Good job typing :+1:

File: tests/plugins/chart/basic_chart.test.ts [hunk 19]
```diff
@@ -1527,24 +1534,37 @@ describe("Chart design configuration", () => {
...
-        datasetIndex: 0,
-        index: 0,
-      };
-    }
+  function getTooltipLabel(
+    chart: ChartJSRuntime,
+    dataset: ChartDataset<
+      keyof ChartTypeRegistry,
+      (number | [number, number] | Point | BubbleDataPoint | null)[]
+    >,
```
Comment (hokolomopo) [inline tests/plugins/chart/basic_chart.test.ts:1542]: This argument seems useless, you can deduce it from chart and datasetIndex

chart.chartJsConfig.data!.datasets[datasetIndex]

**Review by hokolomopo — COMMENTED** [review]
Some details and it should be good :)

**Comment by aath-odoo** [comment]
> Some details and it should be good :)

Updated with the suggested changes, good to go :)

**Review by hokolomopo — COMMENTED** [review]
@robodoo r+
